cmd/bnsd/x/blueaccount/handler.go

```diff
@@ -503,3 +509,55 @@ func (h *deleteAccountHandler) validate(ctx weave.Context, db weave.KVStore, tx
 	}
 	return &msg, nil
 }
+
+type flushDomainHandler struct {
+	auth     x.Authenticator
+	domains  orm.ModelBucket
+	accounts orm.ModelBucket
+}
+
+func (h *flushDomainHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return nil, err
+	}
+	return &weave.CheckResult{GasAllocated: 0}, nil
+}
+
+func (h *flushDomainHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+	// This might be an expensive operation.
+	accountKeys, err := itemKeys(DomainAccounts(db, msg.Domain))
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot list accounts")
+	}
+
+	rootAccountKeySuffix := []byte(":" + msg.Domain + "*")
+	for _, key := range accountKeys {
+		// No name account cannot be deleted as long as the domain exists.
+		if bytes.HasSuffix(key, rootAccountKeySuffix) {
+			continue
+		}
+		if err := db.Delete(key); err != nil {
+			return nil, errors.Wrap(err, "cannot delete an account")
+		}
+	}
+	return &weave.DeliverResult{Data: nil}, nil
+}
+
+func (h *flushDomainHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*FlushDomainMsg, error) {
+	var msg FlushDomainMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+	var domain Domain
+	if err := h.domains.One(db, []byte(msg.Domain), &domain); err != nil {
+		return nil, errors.Wrap(err, "cannot get domain")
+	}
+	if !h.auth.HasAddress(ctx, domain.Owner) {
```

Comment:
I am using `MainSigner` to check the owner's signature.
```go
// MainSigner returns the first permission if any, otherwise nil
func MainSigner(ctx weave.Context, auth Authenticator) weave.Condition {
	signers := auth.GetConditions(ctx)
	if len(signers) == 0 {
		return nil
	}
	return signers[0]
}
```
Can you explain what leads you to use `h.auth.HasAddress`?

Reply:
[`x.Authenticator`](https://github.com/iov-one/weave/blob/master/x/auth.go#L11-L17) is the way to go with any authentication logic. You should use `authenticator.HasAddress` whenever you want to ensure that a certain address signed the transaction. A transaction can be signed multiple times by many parties. Not necessary multisig.

[`x.MainSigner` returns at most a single condition](https://github.com/iov-one/weave/blob/2d99f537f32824a72b22da120e8bc066d6339359/x/auth.go#L66), which is not enough and is not strictly authentication functionality. This should be used only if you need a signer. Usually this is used as a default address fallback to provide a nicer API. For example, when you [create an escrow](https://github.com/iov-one/weave/blob/bcc64032b813241a42fd5b25d46e234f6ca3d885/x/escrow/handler.go#L72), you don't want to give your address explicitly. And there is no need for this, because you have signed the transaction so we know who you are.